src/serac/numerics/functional/functional.hpp

```diff
@@ -413,144 +554,68 @@ class Functional<test(trial), exec> : public mfem::Operator {
+    /// @brief shallow copy of the test space from the associated Functional
+    mfem::ParFiniteElementSpace* test_space_;
 
-    output_T.HostReadWrite();
-    for (int i = 0; i < ess_tdof_list_.Size(); i++) {
-      if constexpr (op == Operation::Mult) {
-        output_T(ess_tdof_list_[i]) = 0.0;
-      }
+    /// @brief shallow copy of the trial space from the associated Functional
+    mfem::ParFiniteElementSpace* trial_space_;
 
-      if constexpr (op == Operation::GradientMult) {
-        output_T(ess_tdof_list_[i]) = input_T(ess_tdof_list_[i]);
-      }
-    }
-  }
+    /// @brief storage for computing the action-of-gradient output
+    mfem::Vector df_;
+  };
 
-  /**
-   * @brief The input set of local DOF values (i.e., on the current rank)
-   */
-  mutable mfem::Vector input_L_;
+  /// @brief The input set of local DOF values (i.e., on the current rank)
+  mutable mfem::Vector input_L_[num_trial_spaces];
 
-  /**
-   * @brief The output set of local DOF values (i.e., on the current rank)
-   */
+  /// @brief The output set of local DOF values (i.e., on the current rank)
   mutable mfem::Vector output_L_;
 
-  /**
-   * @brief The input set of per-element DOF values
-   */
-  mutable mfem::Vector input_E_;
+  /// @brief The input set of per-element DOF values
+  mutable std::array<mfem::Vector, num_trial_spaces> input_E_;
 
-  /**
-   * @brief The output set of per-element DOF values
-   */
+  /// @brief The output set of per-element DOF values
   mutable mfem::Vector output_E_;
 
-  /**
-   * @brief The input set of per-boundaryelement DOF values
-   */
-  mutable mfem::Vector input_E_boundary_;
+  /// @brief The input set of per-boundaryelement DOF values
+  mutable std::array<mfem::Vector, num_trial_spaces> input_E_boundary_;
 
-  /**
-   * @brief The output set of per-boundary-element DOF values
-   */
+  /// @brief The output set of per-boundary-element DOF values
   mutable mfem::Vector output_E_boundary_;
 
-  /**
-   * @brief The output set of local DOF values (i.e., on the current rank) from boundary elements
-   */
+  /// @brief The output set of local DOF values (i.e., on the current rank) from boundary elements
   mutable mfem::Vector output_L_boundary_;
 
-  /**
-   * @brief The set of true DOF values, used as a scratchpad for @p operator()
-   */
-  mutable mfem::Vector my_output_T_;
+  /// @brief The set of true DOF values, a reference to this member is returned by @p operator()
+  mutable mfem::Vector output_T_;
 
-  /**
-   * @brief A working vector for @p GetGradient
-   */
-  mutable mfem::Vector dummy_;
-
-  /**
-   * @brief Manages DOFs for the test space
-   */
+  /// @brief Manages DOFs for the test space
   mfem::ParFiniteElementSpace* test_space_;
 
-  /**
-   * @brief Manages DOFs for the trial space
-   */
-  mfem::ParFiniteElementSpace* trial_space_;
+  /// @brief Manages DOFs for the trial space
+  std::array<mfem::ParFiniteElementSpace*, num_trial_spaces> trial_space_;
```

Comment:
I can see that this is the reason for pointers, although we could use reference wrappers.